lib/puppet/http/session.rb

```diff
@@ -11,29 +12,21 @@ def route_to(name, ssl_context: nil)
     cached = @resolved_services[name]
     return cached if cached
 
-    errors = []
-
     @resolvers.each do |resolver|
       Puppet.debug("Resolving service '#{name}' using #{resolver.class}")
-      resolver.resolve(self, name) do |service|
-        begin
-          service.connect(ssl_context: ssl_context)
-          @resolved_services[name] = service
-          Puppet.debug("Resolved service '#{name}' to #{service.url}")
-          return service
-        rescue Puppet::HTTP::ConnectionError => e
-          errors << e
-          Puppet.debug("Connection to #{service.url} failed, trying next route: #{e.message}")
-        end
+      service = resolver.resolve(self, name, ssl_context: ssl_context)
+      if service
+        @resolved_services[name] = service
+        Puppet.debug("Resolved service '#{name}' to #{service.url}")
+        return service
       end
     end
 
-    errors.each { |e| Puppet.log_exception(e) }
-
+    @resolution_exceptions.each { |e| Puppet.log_exception(e) }
     raise Puppet::HTTP::RouteError, "No more routes to #{name}"
   end
 
-  def create_service(name, server = nil, port = nil)
-    Puppet::HTTP::Service.create_service(@client, name, server, port)
+  def add_exception(exception)
+    @resolution_exceptions << exception
   end
```

Comment:
One problem with using the session to store exceptions is we may resolve multiple routes in one session. So either the `Session#route_to` method needs to clear the `resolutions_exceptions` array before trying to resolve:

```ruby
@resolution_exceptions = []
@resolvers.each do |resolver|
  ...
end
@resolution_exceptions.each { |e| Puppet.log_exception(e) }
```

Or the session could pass a local array into each resolver, and each resolver could add exceptions to the array:

```ruby
resolution_exceptions = []
@resolvers.each do |resolver|
  resolver.resolve(self, name, resolution_exceptions) do |service|
    ...
  end
end
resolution_exceptions.each { |e| Puppet.log_exception(e) }
```

Or the session could pass a callback to be invoked whenever an exception oduring resolution:

```ruby
resolution_exceptions = []
exception_handler = e -> { resolution_exceptions << e }
@resolvers.each do |resolver|
  resolver.resolve(self, name, on_exception: exception_handler) do |service|
    ...
  end
end
resolution_exceptions.each { |e| Puppet.log_exception(e) }
```

Clearing the array during each `route_to` call seems fine for now. We can adjust later if need be.

Reply:
I've updated this to clear `@resolution_exceotions` before trying to resolve